File: Java/Median-of-two-Sorted-Arrays.java
class Solution {
    /**
     * @param A: An integer array.
     * @param B: An integer array.
     * @return: a double whose format is *.5 or *.0
     */
    public int FindKth(int[] A, int[] B, int startA, int startB, int k) {
        if (startA >= A.length) return B[startB + k - 1];
        if (startB >= B.length) return A[startA + k - 1];
        if (k == 1) return Math.min(A[startA], B[startB]);
        int A_key = startA + (k / 2) - 1 < A.length ? A[startA + (k / 2) - 1] : Integer.MAX_VALUE;
        int B_key = startB + (k / 2) - 1 < B.length ? B[startB + (k / 2) - 1] : Integer.MAX_VALUE;
        if (A_key < B_key) return FindKth(A, B, startA + k / 2, startB, k - k / 2);
        else return FindKth(A, B, startA, startB + k / 2, k - k / 2);
    }
    public double findMedianSortedArrays(int[] A, int[] B) {
        // write your code here
        int len = A.length + B.length;
        if ((len & 1) == 0) {
            return (FindKth(A, B, 0, 0, len / 2) + FindKth(A, B, 0, 0, len/ 2 + 1)) / 2.0;
        } else {
            return FindKth(A, B, 0, 0, len / 2 + 1);
        }
    }
}

File: Cpp/Median-of-two-Sorted-Arrays.cpp
#include <vector>

class Solution {
public:
    /**
     * @param A: An integer array.
     * @param B: An integer array.
     * @return: a double whose format is *.5 or *.0
     */
    double findKth(vector<int> &A, vector<int> &B, int leftA, int lenA, int leftB, int lenB, int k) {
        if (lenA > lenB) return findKth(B, A, leftB, lenB, leftA, lenA, k);
        if (lenA == 0) return B[leftB + k - 1];
        if (k == 1) return min(A[leftA], B[leftB]);
        int pa = min(lenA, k / 2), pb = k - pa;
        if (A[leftA + pa - 1] < B[leftB + pb - 1]) {
            return findKth(A, B, leftA + pa, lenA - pa, leftB, lenB, k - pa);
        } else if (A[leftA + pa - 1] > B[leftB + pb - 1]) {
            return findKth(A, B, leftA, lenA, leftB + pb, lenB - pb, k - pb);
        } else {
            return A[leftA + pa - 1];
        }
    }
    double findMedianSortedArrays(vector<int> &A, vector<int> &B) {
        // write your code here
        int len = A.size() + B.size();
        if (len & 0x1) return findKth(A, B, 0, A.size(), 0, B.size(), len / 2 + 1);
        else return (findKth(A, B, 0, A.size(), 0, B.size(), len / 2)
            + findKth(A, B, 0, A.size(), 0, B.size(), len / 2 + 1)) / 2;
    }
};

